src/ray/gcs/gcs_client/gcs_client.cc

```diff
@@ -139,5 +139,286 @@ std::pair<std::string, int> GcsClient::GetGcsServerAddress() const {
+  }
+  return ConvertGrpcStatus(status);
+}
+
+Status GcsSyncClient::InternalKVKeys(const std::string &ns,
+                                     const std::string &prefix,
+                                     int64_t timeout_ms,
+                                     std::vector<std::string> &results) {
+  grpc::ClientContext context;
+  GrpcClientContextWithTimeoutMs(context, timeout_ms);
+
+  rpc::InternalKVKeysRequest request;
+  request.set_namespace_(ns);
+  request.set_prefix(prefix);
+
+  rpc::InternalKVKeysReply reply;
+
+  grpc::Status status = kv_stub_->InternalKVKeys(&context, request, &reply);
+  if (status.ok()) {
+    if (reply.status().code() == static_cast<int>(StatusCode::OK)) {
+      results = std::vector<std::string>(reply.results().begin(), reply.results().end());
+      return Status::OK();
+    }
+    return HandleGrpcError(reply.status());
+  }
+  return ConvertGrpcStatus(status);
+}
+
+Status GcsSyncClient::InternalKVExists(const std::string &ns,
+                                       const std::string &key,
+                                       int64_t timeout_ms,
+                                       bool &exists) {
+  grpc::ClientContext context;
+  GrpcClientContextWithTimeoutMs(context, timeout_ms);
+
+  rpc::InternalKVExistsRequest request;
+  request.set_namespace_(ns);
+  request.set_key(key);
+
+  rpc::InternalKVExistsReply reply;
+
+  grpc::Status status = kv_stub_->InternalKVExists(&context, request, &reply);
+  if (status.ok()) {
+    if (reply.status().code() == static_cast<int>(StatusCode::OK)) {
+      exists = reply.exists();
+      return Status::OK();
+    }
+    return HandleGrpcError(reply.status());
+  }
+  return ConvertGrpcStatus(status);
+}
+
+Status GcsSyncClient::PinRuntimeEnvUri(const std::string &uri,
+                                       int expiration_s,
+                                       int64_t timeout_ms) {
+  grpc::ClientContext context;
+  GrpcClientContextWithTimeoutMs(context, timeout_ms);
+
+  rpc::PinRuntimeEnvURIRequest request;
+  request.set_uri(uri);
+  request.set_expiration_s(expiration_s);
+
+  rpc::PinRuntimeEnvURIReply reply;
+
+  grpc::Status status = runtime_env_stub_->PinRuntimeEnvURI(&context, request, &reply);
+  if (status.ok()) {
+    if (reply.status().code() == static_cast<int>(StatusCode::OK)) {
+      return Status::OK();
+    } else if (reply.status().code() == static_cast<int>(StatusCode::GrpcUnavailable)) {
+      std::string msg =
+          "Failed to pin URI reference for " + uri + " due to the GCS being " +
+          "unavailable, most likely it has crashed: " + reply.status().message() + ".";
+      return Status::GrpcUnavailable(msg);
+    }
+    std::string msg = "Failed to pin URI reference for " + uri +
+                      " due to unexpected error " + reply.status().message() + ".";
+    return Status::GrpcUnknown(msg);
+  }
+  return ConvertGrpcStatus(status);
+}
+
+Status GcsSyncClient::GetAllNodeInfo(int64_t timeout_ms,
```

Comment:
Hmm it is pretty surprising that our default timeout was None for all of these calls... I guess it doesn't need to be done in this PR, but we may need to add timeout for these. cc @scv119

Reply:
Yeah, I'm trying to not modify the behavior with this PR, let's add that later :)